File: Algorithms on Strings/Programming-Assignment-4/suffix_array_long/suffix_array_long.cpp
/**
 * Algorithms on Strings/Week 4/Problem 2
 * @file suffix_array_long.cpp
 * @author your name (you@domain.com)
 * @brief
 * @version 0.1
 * @date 2022-02-23
 *
 * @copyright Copyright (c) 2022
 *
 */

#include <bits/stdc++.h>

using namespace std;

const int SIZE_OF_ALPHABET = 5;

int getIndex(char c)
{
  switch (c)
  {
  case 'A':
    return 1;
  case 'C':
    return 2;
  case 'G':
    return 3;
  case 'T':
    return 4;
  default:
    return 0;
  }
}

vector<int> sortDoubled(const string &s, int L, const vector<int> &order, const vector<int> &_class)
{
  vector<int> count(s.size());
  vector<int> new_order(s.size());

  for (int i = 0; i < s.size(); ++i)
  {
    ++count[_class[i]];
  }

  for (int j = 1; j < s.size(); ++j)
  {
    count[j] += count[j - 1];
  }

  for (int i = s.size() - 1; i >= 0; --i)
  {
    int start = (order[i] - L + s.size()) % s.size();
    int cl = _class[start];
    new_order[--count[cl]] = start;
  }

  return new_order;
}

vector<int> ComputeCharClasses(const string &s, const vector<int> &order)
{
  vector<int> _class(s.size());

  for (int i = 1; i < s.size(); ++i)
  {
    if (s[order[i]] != s[order[i - 1]])
    {
      _class[order[i]] = _class[order[i - 1]] + 1;
    }
    else
    {
      _class[order[i]] = _class[order[i - 1]];
    }
  }

  return _class;
}

vector<int> SortCharacters(const string &s)
{
  vector<int> order(s.size());
  vector<int> count(SIZE_OF_ALPHABET);

  for (auto &ch : s)
  {
    ++count[getIndex(ch)];
  }

  for (int j = 1; j < SIZE_OF_ALPHABET; ++j)
  {
    count[j] += count[j - 1];
  }

  for (int i = s.size() - 1; i >= 0; --i)
  {
    char c = s[i];
    count[getIndex(c)] -= 1;
    order[count[getIndex(c)]] = i;
  }
  return order;
}

vector<int> updateClasses(const vector<int> &new_order, const vector<int> &_class, int L)
{
  const int n = new_order.size();
  vector<int> new_class(n);

  for (int i = 1; i < n; ++i)
  {
    int cur = new_order[i], prev = new_order[i - 1];
    int mid = cur + L, mid_prev = (prev + L) % n;

    if (_class[cur] != _class[prev] || _class[mid] != _class[mid_prev])
    {
      new_class[cur] = new_class[prev] + 1;
    }
    else
    {
      new_class[cur] = new_class[prev];
    }
  }

  return move(new_class);
}

vector<int> buildSuffixArray(const string &S)
{
  vector<int> _order = SortCharacters(S);
  vector<int> _class = ComputeCharClasses(S, _order);
  int L = 1;

  while (L < S.size())
  {
    _order = sortDoubled(S, L, _order, _class);
    _class = updateClasses(_order, _class, L);
    L *= 2;
  }

  return move(_order);
}

int main()
{
  string s;
  cin >> s;
  vector<int> order = buildSuffixArray(s);

  for (auto elem : order)
  {
    cout << elem << ' ';
  }

  return 0;
}

File: Algorithms on Strings/Programming-Assignment-4/suffix_array_long/suffix_array_long.py
#

def SortCharacters(s):
    order = [0] * len(s)
    count = {'$': 0, "A": 0, 'C': 0, 'G': 0, 'T': 0}
    for char in s:
        count[char] += 1
    symb = ['$', 'A', 'C', 'G', 'T']
    for i in range(1, 5):
        count[symb[i]] += count[symb[i-1]]
    for j in range(len(s) - 1, -1, -1):
        c = s[j]
        count[c] -= 1
        order[count[c]] = j
    return order


def ComputeCharClasses(s, order):
    class_chars = [0] * len(s)
    for i in range(1, len(s)):
        if s[order[i]] == s[order[i-1]]:
            class_chars[order[i]] = class_chars[order[i-1]]
        else:
            class_chars[order[i]] = class_chars[order[i-1]] + 1
    return class_chars


def SortDoubled(s, L, old_order, old_class):
    count = [0] * len(s)
    new_order = [0] * len(s)
    for i in range(len(s)):
        count[old_class[i]] += 1
    for i in range(1, len(s)):
        count[i] += count[i-1]
    for j in range(len(s) - 1, -1, -1):
        start = (old_order[j] - L + len(s)) % len(s)
        cl = old_class[start]
        count[cl] -= 1
        new_order[count[cl]] = start
    return new_order


def UpdateClasses(new_order, old_class, L):
    n = len(new_order)
    new_class = [0] * n
    for i in range(1, n):
        cur = new_order[i]
        mid = (cur + L) % n
        prev = new_order[i-1]
        mid_prev = (prev + L) % n
        if old_class[cur] == old_class[prev] and old_class[mid] == old_class[mid_prev]:
            new_class[cur] = new_class[prev]
        else:
            new_class[cur] = new_class[prev] + 1
    return new_class


def BuildSuffixArray(S):
    order = SortCharacters(S)
    class_ = ComputeCharClasses(S, order)
    L = 1
    while L < len(S):
        order = SortDoubled(S, L, order, class_)
        class_ = UpdateClasses(order, class_, L)
        L = 2 * L
    return order


if __name__ == '__main__':
    text = input()
    suffix_array = BuildSuffixArray(text)
    for elem in suffix_array:
        print(elem, end=' ')